stablehlo/reference/NumPy.cpp

```diff
@@ -0,0 +1,323 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "stablehlo/reference/NumPy.h"
+
+#include <fstream>
+#include <regex>
+
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/Types.h"
+
+namespace mlir {
+namespace stablehlo {
+namespace numpy {
+namespace {
+
+constexpr char kMagicString[] = "\x93NUMPY";
+constexpr int kMagicStringSize = 6;
+constexpr char kMajorVersion = 0x01;
+constexpr char kMinorVersion = 0x00;
+
+template <typename T>
+struct IsComplexT : public std::false_type {};
+template <typename T>
+struct IsComplexT<std::complex<T>> : public std::true_type {};
+
+}  // namespace
+
+// Determine the NumPy type abbreviation for the given underlying data type T.
+template <typename T>
+static constexpr char getNumPyType() {
+  if constexpr (std::is_same_v<T, bool>) return 'b';
+  if constexpr (IsComplexT<T>::value) return 'c';
+  if constexpr (std::is_floating_point_v<T>) return 'f';
+  if constexpr (std::is_integral_v<T>) {
+    if constexpr (std::is_signed_v<T>)
+      return 'i';
+    else
+      return 'u';
+  }
+
+  llvm_unreachable("Unknown type");
+}
+
+// Determine the endianness of the machine.
+static char getEndianChar() {
```

Comment:
Let's use LLVM to determine endianness. This is cool but I've been staring at it confused why it works :)

```
#include "llvm/Support/Endian.h"

llvm::support::endian::system_endianness() ==
      llvm::support::endianness::big
```